### DISCUSSION TITLE
mixed datatype in v1.6.12

### DISCUSSION THREAD

**Discussion by jopabinia** [iqtree/iqtree2#213] [Q&A] [answered]
I have an alignment with 3 ribosomal genes to treat as nucleotides, and 2 protein coding genes as amino acids. It's all concatenated together, so I thought I would use a partitioned analysis. I tried it 2 different ways:

1) in nexus file, 
FORMAT datatype=mixed(nucleotide:1-1486,protein:1487-2101)

2) using a nexus file without the above, but adding a partition file,
#nexus
begin sets;
  charset 18S = 1-514; 
  charset 28S = 515-622; 
  charset 16S = 623-1486; 
  charset cox1 = 1487-1764;
  charset cytb = 1765-2101; 
end;

The first one completely fails, is this a function of using v1.6.12 (which is what's on my cluster)?
The second one won't acknowledge different datatypes, which is fair because it's not specified. But it then just assumes they are all amino acids and that the nucleotide loci are missing most of the characters.

What am I doing wrong? Per this http://www.iqtree.org/doc/Advanced-Tutorial#partitioned-analysis-with-mixed-data
should I not have concatenated the alignments, and use each one separately?

Thanks!

**Comment by roblanf** [comment]
@jopabinia according to the docs (I am not familiar with the internals on this point!) you also need to specify the models. 

E.g. the docs show this:

```
#nexus
begin sets;
    charset part1 = dna.phy: 1-100 201-300;
    charset part2 = dna.phy: 101-200;
    charset part3 = prot.phy: 1-400;
    charset part4 = prot.phy: 401-600;
    charset part5 = codon.phy:CODON, *;
    charpartition mine = HKY:part1, GTR+G:part2, LG+G:part3, WAG+I+G:part4, GY:part5;
end;
```

I'm assuming that the models here are what allows IQ-TREE to know what the datatype is. 

Rob

**Reply by bqminh** [reply]
The line "charpartition" is not mandatory. In the absence of this command, IQ-TREE will use all "charset" as partitions and find the best-fit model for each one of them.
In the presence of "charpartition" command, IQ-TREE will only include those charsets that are listed under charpartition (non-listed charsets will be ignored).

**Reply by roblanf** [reply, marked as answer]
@jopabinia so just to follow up on @bqminh's advice, all you need to do is put your alignments in different files. As long as no one file contains more than one datatype, you're OK. So you could do it like my example above, but without the `charpartition`, i.e. one alignment file for DNA and one for amino acids:

```
#nexus
begin sets;
    charset part1 = dna.phy: 1-100 201-300;
    charset part2 = dna.phy: 101-200;
    charset part3 = prot.phy: 1-400;
    charset part4 = prot.phy: 401-600;
end;
```

or you could just keep each alignment in its own file, something like this:

```
#nexus
begin sets;
charset 18S = 18S.phy: 1-514;
charset 28S = 28S.phy: 1-105;
charset cox1 = cox1.phy 1-500;
end;
```

Rob

**Reply by bqminh** [reply]
@roblanf if each file represents a partition, you can simplify by skipping the position specification and iqtree will assign all columns of that alignment file to the partition, e.g.:

```
#nexus
begin sets;
charset 18S = 18S.phy; [ all sites of 18S.phy will be partition 18S ]
charset 28S = 28S.phy;
charset cox1 = cox1.phy;
end;
```

**Comment by jopabinia** [comment]
Thanks! But, oh dear because modelfinder is what I was hoping to do!

It did work by separating the two partitions and doing modelfinder separately, as I guessed.

Thanks!

**Reply by bqminh** [reply]
Such line in nexus file

```
FORMAT datatype=mixed(nucleotide:1-1486,protein:1487-2101)
```

is not supported in IQ-TREE. It requires that an alignment file can only contain one datatype, i.e., you can't mix DNA and protein in a single alignment file. So you should not concatenate a DNA and a protein alignment, but rather having two separate files and specify the partition file accordingly.
